src/Microsoft.FeatureManagement/FeatureManagerSnapshot.cs

```diff
@@ -12,37 +12,99 @@ namespace Microsoft.FeatureManagement
+            IDynamicFeatureManager dynamicFeatureManager)
         {
             _featureManager = featureManager ?? throw new ArgumentNullException(nameof(featureManager));
+            _dynamicFeatureManager = dynamicFeatureManager ?? throw new ArgumentNullException(nameof(dynamicFeatureManager));
         }
 
-        public async IAsyncEnumerable<string> GetFeatureNamesAsync([EnumeratorCancellation]CancellationToken cancellationToken)
+        public async IAsyncEnumerable<string> GetFeatureFlagNamesAsync([EnumeratorCancellation]CancellationToken cancellationToken)
         {
-            if (_featureNames == null)
+            if (_featureFlagNames == null)
             {
                 var featureNames = new List<string>();
 
-                await foreach (string featureName in _featureManager.GetFeatureNamesAsync(cancellationToken).ConfigureAwait(false))
+                await foreach (string featureName in _featureManager.GetFeatureFlagNamesAsync(cancellationToken).ConfigureAwait(false))
                 {
                     featureNames.Add(featureName);
                 }
 
-                _featureNames = featureNames;
+                _featureFlagNames = featureNames;
             }
 
-            foreach (string featureName in _featureNames)
+            foreach (string featureName in _featureFlagNames)
             {
                 yield return featureName;
             }
         }
 
+        public async IAsyncEnumerable<string> GetDynamicFeatureNamesAsync([EnumeratorCancellation]CancellationToken cancellationToken = default)
+        {
+            if (_dynamicFeatureNames == null)
+            {
+                var dynamicFeatureNames = new List<string>();
+
+                await foreach (string featureName in _featureManager.GetFeatureFlagNamesAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    dynamicFeatureNames.Add(featureName);
+                }
+
+                _dynamicFeatureNames = dynamicFeatureNames;
+            }
+
+            foreach (string featureName in _dynamicFeatureNames)
+            {
+                yield return featureName;
+            }
+        }
+
+        public async ValueTask<T> GetVariantAsync<T>(string feature, CancellationToken cancellationToken)
+        {
+            string cacheKey = GetVariantCacheKey<T>(feature);
+
+            //
+            // First, check local cache
+            if (_flagCache.ContainsKey(feature))
+            {
+                return (T)_variantCache[cacheKey];
+            }
+
+            T variant = await _dynamicFeatureManager.GetVariantAsync<T>(feature, cancellationToken).ConfigureAwait(false);
+
+            _variantCache[cacheKey] = variant;
+
+            return variant;
+        }
+
+        public async ValueTask<T> GetVariantAsync<T, TContext>(string feature, TContext context, CancellationToken cancellationToken)
+        {
+            string cacheKey = GetVariantCacheKey<T>(feature);
```

Comment:
Should the context be part of the cache key in this case too?

Reply:
We didn't factor context into the snapshot for feature flags. There was a discussion for this in this issue: https://github.com/microsoft/FeatureManagement-Dotnet/issues/87

Reply:
Indeed. That carries some complexity. Let's see if it's brought up again in the future. Leave it as-is for now.